_posts/2018-01-11-keystone-v3.md

```diff
@@ -58,23 +58,27 @@ KeystoneApi keystone = ContextBuilder.newBuilder("openstack-keystone-3")
+In the Keystone V3 API, most of the update operations are carried out by sending `PATCH` HTTP requests. The `PATCH` verb, however, is not supported by jclouds' default Java HTTP driver. If you plan to use such API methods, you will use an HTTP driver with support for `PATCH`, such as the [OkHttp](https://github.com/jclouds/jclouds/tree/master/drivers/okhttp) or [ApacheHC](https://github.com/jclouds/jclouds/tree/master/drivers/apachehc) drivers.
+
+To configure an HTTP driver, add the corresponding module to the list of modules passed to the `ContextBuilder` when creating your jclouds context. For example:
 
 {% highlight java %}
 KeystoneApi keystone = ContextBuilder.newBuilder("openstack-keystone-3")
    .endpoint("http://openstack-keystone/identity/v3")
    .credentials("domain:admin", "password")
    .overrides(overrides)
-   .modules(ImmutableSet.of(new SLF4JLoggingModule(), new OkHttpCommandExecutorServiceModule()))
+   .modules(ImmutableSet.of(new SLF4JLoggingModule(), new OkHttpCommandExecutorServiceModule())) // use OkHttp driver
    .buildApi(KeystoneApi.class);
 {% endhighlight %}
 
-# Upgrade notes and breaking changes
+# Notes and breaking changes
 
-In order to support V2 and V3, a major refactor has been done to the `openstack-keystone` API and many packages and classes have been renamed, moved and deleted. If your code is relying on constants or other global classes, you may need to update the package references.
+Supporting both the V2 and V3 Keystone APIs required a major refactor of the `openstack-keystone` API. Many packages and classes have been renamed, moved and deleted as a result. If your code uses constants or other global classes, you may need to update the following package references:
 
-* Class `KeystoneProperties` has been moved to package `org.jclouds.openstack.keystone.config`.
-* Class `CredentialTypes` has been moved to package `org.jclouds.openstack.keystone.auth.config`.
+* Class `KeystoneProperties` has been moved to `org.jclouds.openstack.keystone.config`.
+* Class `CredentialTypes` has been moved to `org.jclouds.openstack.keystone.auth.config`.
 * The `KeystoneAuthenticationModule` and the `AuthenticationApiModule` have been refactored and generalised into:
   * `AuthenticationModule` - Providing authentication services to all OpenStack APIs and providers.
   * `ServiceCatalogModule` - Providing endpoint resolution to all OpenStack APIs and providers.
+
+Replace uses of `KeystoneAuthenticationModule` with `AuthenticationModule`, and uses of `AuthenticationApiModule` with `ServiceCatalogModule`.
```

Comment:
This is not strictly true, and that's why I wrote it that way. The KeystoneAuthModule contained both, the authentication configuration and the service catalog one, and the AuthenticationApiModule just bound the internal Authentication API.

I've refactored it in a way that makes more sense, moving all config to the AuthenticationModule and the service catalog parsing to the Service Catalog module, but I don't think we need to mention that. Those modules are already included by all OpenStack APIs, so that info is interesting only for those writing new APIs. I wanted to mention the refactoring, but users won't need to replace anything when upgrading, so I'd just remove this last sentence.

Apart from this, LGTM. Many thanks for reviewing! :)

Reply:
> Those modules are already included by all OpenStack APIs, so that info is interesting only for those writing new APIs.

So is the guidance something like "Use `AuthenticationModule` and `ServiceCatalogModule` when developing OpenStack APIs."?

Reply:
Yes, that's fine :)